### PR TITLE
Problem: unchecked dirty bytes and missing length overflow checks

### PR DESCRIPTION
**Pull request by yihuang**
# go-abi Polish — Plan & Execution Summary

## Overview

Audit of the `github.com/yihuang/go-abi` project for discrepancies against the ABI specification,
code quality issues, and missing test coverage. All changes were developed test-first (TDD).

---

### Bug 1: FixedBytes Encode Leaves Garbage in Padding

**Files changed:** `generator/encoders.go`, `stdlib.abi.go`, `stdlib_uint256.abi.go`

**Problem:** `genFixedBytesEncoding` wrote only `t.Size` bytes into the 32-byte ABI word slot
(e.g., `EncodeBytes3` wrote 3 bytes, left bytes 4–31 untouched). When `EncodeTo` was called
with a reused (non-zeroed) buffer, the padding region contained garbage. `DecodeBytesN`
would then reject it with `ErrDirtyPadding`, breaking round-trips via `EncodeTo`.

**Fix:** Zero the remaining `(32 - t.Size)` padding bytes after writing the value.
Return `32` (full ABI word size) instead of `t.Size`.

**Test:** `TestFixedBytesEncodePadding` — 5 sub-tests (`bytes1`, `bytes3`, `bytes7`, `bytes15`, `bytes31`)
fill a buffer with `0xFF` poison bytes, call `EncodeBytesN`, verify padding region is zeroed.

### Bug 2: FixedBytes Decode Returns Wrong Consumed Count

**Files changed:** `generator/decoders.go`, `stdlib.abi.go`, `stdlib_uint256.abi.go`

**Problem:** `DecodeBytesN` returned `t.Size` instead of `32`. Per the ABI spec, all fixed bytes
types (`bytes1` through `bytes32`) occupy a full 32-byte word. Also missing a `len(data) < 32` length guard.

**Fix:** Return `32` and add `if len(data) < 32 { return ... io.ErrUnexpectedEOF }`.

**Test:** `TestFixedBytesDecodePadding` — 4 variants × clean/dirty input verifies correct consumed
count and padding validation.

### Bug 3: DecodeAddress Panics on Short Input

**Files changed:** `generator/decoders.go`, `stdlib.abi.go`, `stdlib_uint256.abi.go`

**Problem:** `genAddressDecoding` accessed `data[12:32]` without checking `len(data) < 32`,
causing a slice bounds panic on short input.

**Fix:** Add `if len(data) < 32 { return common.Address{}, 0, io.ErrUnexpectedEOF }`.

**Test:** `TestDecodeAddressValidation` — sub-tests for short data (2 bytes, 31 bytes) verify
`io.ErrUnexpectedEOF` is returned instead of a panic.

### Bug 4: Packed Big.Int Encode/Decode for >64-bit Types

**Files changed:** `generator/encoders.go`, `generator/decoders.go`, `stdlib.abi.go`, `stdlib_uint256.abi.go`

**Problem:** `genPackedIntEncoding` passed `buf[:byteSize]` (e.g. 16 bytes) to `EncodeBigInt`,
which internally does `n.FillBytes(buf[32-l:32])` requiring a 32-byte buffer → panic.
Similarly, `genPackedIntDecoding` passed `data[:byteSize]` to `DecodeBigInt` which reads
`data[:32]` → out-of-bounds.

**Fix:** Use a temporary `[32]byte` buffer for `EncodeBigInt`, then copy the needed bytes.
For decoding, extend packed bytes to 32 via sign-extension (signed) or zero-padding (unsigned)
before calling `DecodeBigInt`.

**Test:** `TestPackedLargeInts` — round-trips `uint128` and `int128` through packed encode/decode.

### Bug 5: EmptyIndexed Event Missing Event Signature Topic

**Files changed:** `generator/generator.go`, `tests/test.abi.go`, `tests/test_uint256.abi.go`

**Problem:** Events with zero indexed fields embedded `abi.EmptyIndexed`, whose `EncodeTopics()`
returned `nil`. Per ABI spec, `topics[0]` must always be the event signature hash.

**Fix:** Generate a custom `EncodeTopics` method for empty-indexed events that returns
`[]common.Hash{EventTopic}`.

**Test:** `TestEventEmptyIndexed` — verifies `len(topics) == 1` and `topics[0]` is the event sig.

---

## Additional Tests Added (All Passing)

| Test | File | What it validates |
|------|------|-------------------|
| `TestFixedBytesEncodePadding` | `utils_test.go` | EncodeBytesN zeros the padding region (5 variants) |
| `TestFixedBytesDecodePadding` | `utils_test.go` | DecodeBytesN returns 32 and validates padding (4 variants × clean/dirty) |
| `TestDecodeUintPadding` | `utils_test.go` | Dirty high bytes rejected in unsigned decode |
| `TestDecodeIntSignExtension` | `utils_test.go` | Partial sign extension rejected; proper 256-bit accepted |
| `TestDecodeBoolValidation` | `utils_test.go` | Only 0x00/0x01 accepted; 0x02, 0xFF, dirty high bytes rejected |
| `TestDecodeAddressValidation` | `utils_test.go` | First 12 bytes zero; short input returns error not panic |
| `TestDecodeBigIntNegative` | `utils_test.go` | Negative int256 round-trip; negative-as-unsigned rejected |
| `TestEventDynamicIndexed` | `tests/abi_test.go` | Keccak256 hash of ABI-encoded string in topic |
| `TestEventEmptyIndexed` | `tests/abi_test.go` | Empty indexed fields: topic[0] = event sig only |
| `TestPackedLargeInts` | `tests/packed_test.go` | Packed uint128/int128 round-trip |
| `TestPackedFixedArrayStruct` | `tests/packed_test.go` | Packed struct with `address[3]` + `uint256[2]` round-trip |
| `TestParseHumanReadableABI_AddressPayable` | `human_test.go` | `address payable` normalized to `address` |
| Circular struct errors | `human_test.go` | A→B→A and self-referencing Node produce errors |

---

## Code Quality Improvements

| Area | File(s) | Change |
|------|---------|--------|
| **Receiver consistency** | `types.go` | `EmptyTuple`/`EmptyIndexed` methods use consistent value receivers |
| **Dead code** | `human.go` | `functionRegex` simplified — unused capture groups removed |
| **Inefficient helper** | `generator/external_tuples_test.go` | Custom recursive `contains()` replaced with `strings.Contains` |
| **Duplicate assertions** | `generator/alias_import_test.go` | Redundant second assertion block removed |
| **Testify imports** | `human_test.go`, `tests/nested_tuples_test.go`, `go.mod` | Consolidated from dual `stretchr`+`test-go` to single `test-go/testify` |
| **CI** | `.github/workflows/go.yml` | Added `go vet` step for both build tags |
| **`address payable`** | `human.go` | Strip ` payable` qualifier before parameter parsing |
| **Packed >64-bit int** | `generator/encoders.go`, `generator/decoders.go` | Temp 32-byte buffer for `EncodeBigInt`+`DecodeBigInt` calls |
| **EmptyIndexed topics** | `generator/generator.go` | Generate `EncodeTopics` returning `[eventSig]` for zero-indexed events |
| **Artifact tracking** | `.gitignore` | Added patterns for coverage/`coverage/`, `*.prof`, `cov.out`, `cpu.out`, `uint256.out`, `out.prof`, `tests.test`, `docs/abi-spec.rst` |

---

## Commit History

```
575026e feat: complete polish - remaining test coverage, CI vet step, and bugfixes
9bc7b76 chore: gitignore untracked docs/abi-spec.rst and test artifacts
338cacd refactor: improve code consistency and remove dead code
138b61e chore: update .gitignore to exclude coverage artifacts and profiling output
904ae49 fix: three ABI encoding/decoding bugs in generated code
12eb212 test: add regression tests exposing fixed bytes padding bugs, DecodeAddress panic, and validation edge cases
```

---

## Original Plan — Completion Status

| # | Item | Status |
|---|------|--------|
| 1 | FixedBytes encode padding bug | ✅ Fixed |
| 2 | DecodeString/DecodeBytes overflow | ✅ Already handled by existing test |
| 3 | `functionRegex` unused capture groups | ✅ Simplified |
| 4 | `parseLineWithStructs` error discarding | ✅ Inspected — not a bug |
| 5 | `EmptyTuple`/`EmptyIndexed` receivers | ✅ Made consistent |
| 6 | `EncodeBigInt` mutation | ✅ Inspected — no issue |
| 7 | Duplicate testify imports | ✅ Consolidated to `test-go/testify` |
| 8 | `TestAliasImports` duplicate assertions | ✅ Removed |
| 9a | Anonymous event test | ↪️ Needs JSON ABI (human parser hardcodes `anonymous:false`) |
| 9b | `address payable` normalization | ✅ Fixed + tested |
| 9c | Padding validation edge cases | ✅ Done (DecodeUint, DecodeInt, DecodeBool, DecodeAddress) |
| 9d | Dynamic indexed field event test | ✅ Done (keccak256 hash verification) |
| 9e | Recursive struct detection | ✅ Tested |
| 9f | Negative big.Int roundtrip | ✅ Done |
| 9g | Boundary value tests | ✅ Already covered by existing tests |
| 9h | Packed extended tests | ✅ Done (uint128, int128, fixed arrays) |
| 10 | Simplify `contains()` | ✅ Replaced with `strings.Contains` |
| 11 | Missing uint136-248 constants | ✅ Intentional (>64-bit uses `*big.Int`) |
| 12 | CI `go vet` | ✅ Added to workflow |

---

## Regression Check

- `go build ./...` — ✅ clean (both `!uint256` and `uint256` tags)
- `go vet ./...` — ✅ clean
- `go test ./... -count=1` — ✅ all 4 packages pass
- Benchmarks — ✅ no performance regression

### CODE REVIEW

File: generator/encoders.go [hunk 1]
```diff
@@ -126,7 +126,13 @@ func (g *Generator) genBytesEncoding() {
 // genFixedBytesEncoding generates encoding for fixed bytes types
 func (g *Generator) genFixedBytesEncoding(t ethabi.Type) {
 	g.L("\tcopy(buf[:%d], value[:])", t.Size)
-	g.L("\treturn %d, nil", t.Size)
+	if t.Size < 32 {
+		// Zero out padding bytes to ensure clean ABI 32-byte slot
+		g.L("\tfor i := %d; i < 32; i++ {", t.Size)
+		g.L("\t\tbuf[i] = 0")
+		g.L("\t}")
```
Comment (mmsqe) [inline generator/encoders.go]:
```suggestion
		g.L("\tclear(buf[%d:32])", t.Size)
```

**Review by mmsqe — COMMENTED** [review]
seems 338cacd allows malformed signature now

**Comment by yihuang** [comment]
> seems [338cacd](https://github.com/yihuang/go-abi/commit/338cacd8a9121739391577686155ffda7a9bf375) allows malformed signature now

reverted function regex change.

File: generator/encoders.go [hunk 2]
```diff
@@ -126,7 +126,11 @@ func (g *Generator) genBytesEncoding() {
 // genFixedBytesEncoding generates encoding for fixed bytes types
 func (g *Generator) genFixedBytesEncoding(t ethabi.Type) {
 	g.L("\tcopy(buf[:%d], value[:])", t.Size)
-	g.L("\treturn %d, nil", t.Size)
+	if t.Size < 32 {
+		// Zero out padding bytes to ensure clean ABI 32-byte slot
+		g.L("\tclear(buf[%d:32])", t.Size)
+	}
+	g.L("\treturn 32, nil")
```
[the remark this answers is not in the document]
Comment (yihuang) [reply generator/encoders.go:133]: Fixed-size buffers are pre-calculated, the buffer is assumed to be big enough.

**Review by mmsqe — APPROVED** [review]
